pkg/detectors/gcp/gcp.go

```diff
@@ -90,41 +103,144 @@ func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (result
-				continue
-			}
-			if credentials != nil {
-				_, err = credentials.TokenSource.Token()
-				if err == nil {
-					s.Verified = true
-				}
-			}
+			client := s.getClient()
+			isVerified, verificationErr := isValidGCPServiceAccountKey(ctx, client, creds)
+			s1.Verified = isVerified
+			s1.SetVerificationError(verificationErr, creds.ClientEmail)
 		}
 
-		results = append(results, s)
+		results = append(results, s1)
 	}
 
 	return
 }
 
-func (s Scanner) IsFalsePositive(_ detectors.Result) bool {
-	return false
+// isValidGCPServiceAccountKey checks if the provided GCP service account key is valid. It verifies this by
+// comparing the public key extracted from the certificate retrieved from the `ClientX509CertURL`
+// with the public key derived from the `PrivateKey` field of the `gcpKey` struct. If the public keys
+// match, the function returns true, indicating a valid service account key. Otherwise, it returns false.
+//
+// Note: If the service account is expired or disabled, the request to fetch the certificate from the
+// `ClientX509CertURL` will fail with a 404 status code. In this case, the function will return false,
+// indicating that the service account key is invalid, regardless of the public key comparison.
+func isValidGCPServiceAccountKey(ctx context.Context, client *http.Client, key gcpKey) (bool, error) {
+	certPEM, ok, err := fetchCert(ctx, client, key.ClientX509CertURL, key.PrivateKeyID)
+	if err != nil {
+		return false, err
+	}
+	if !ok {
+		return false, nil
+	}
+
+	publicKey, ok := extractPublicKeyFromCert(certPEM)
+	if !ok {
+		return false, nil
+	}
+
+	privateKey, ok := extractPrivateKey(key.PrivateKey)
+	if !ok {
+		return false, nil
+	}
+
+	k, ok := privateKey.Public().(*rsa.PublicKey)
+	if !ok {
+		return false, nil
+	}
+
+	return publicKey.Equal(k), nil
+}
+
+// fetchCert downloads the certificate from the specified URL and returns its PEM-encoded string.
+func fetchCert(ctx context.Context, client *http.Client, url, privKeyID string) (string, bool, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return "", false, err
+	}
+
+	res, err := client.Do(req)
+	if err != nil {
+		return "", false, err
+	}
+	defer func() {
+		_, _ = io.Copy(io.Discard, res.Body)
+		_ = res.Body.Close()
+	}()
+
+	switch {
+	case res.StatusCode >= http.StatusOK && res.StatusCode < http.StatusMultipleChoices:
+		var resp map[string]string
+		if err := json.NewDecoder(res.Body).Decode(&resp); err != nil {
+			return "", false, err
+		}
+
+		cert, ok := resp[privKeyID]
+		return cert, ok, nil
+	case res.StatusCode == http.StatusNotFound:
+		// The service account is expired or disabled.
+		return "", false, nil
+	default:
+		return "", false, fmt.Errorf("unexpected HTTP response status %d", res.StatusCode)
+	}
+
+}
+
+// extractPublicKeyFromCert decodes the PEM-encoded certificate to extract the RSA public key.
+func extractPublicKeyFromCert(certPEM string) (*rsa.PublicKey, bool) {
```

Comment:
Is it worth handling these errors? The input is coming from a trusted API in a known format; I'd think that any failures would be highly unusual and something worth looking at.